src/clj/puppetlabs/services/jruby/jruby_puppet_schemas.clj

```diff
@@ -0,0 +1,164 @@
+(ns puppetlabs.services.jruby.jruby-puppet-schemas
+  (:require [schema.core :as schema]
+            [puppetlabs.services.jruby.puppet-environments :as puppet-env])
+  (:import (java.util.concurrent BlockingDeque)
+           (clojure.lang Atom Agent IFn)
+           (com.puppetlabs.puppetserver PuppetProfiler JRubyPuppet EnvironmentRegistry)
+           (org.jruby.embed ScriptingContainer)))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;; Schemas
+
+(def pool-queue-type
+  "The Java datastructure type used to store JRubyPuppet instances which are
+  free to be borrowed."
+  BlockingDeque)
+
+(defrecord PoisonPill
+  ;; A sentinel object to put into a pool in case an error occurs while we're trying
+  ;; to populate it.  This can be used by the `borrow` functions to detect error
+  ;; state in a thread-safe manner.
+  [err])
+
+(defrecord RetryPoisonPill
+  ;; A sentinel object to put into an old pool when we swap in a new pool.
+  ;; This can be used to build `borrow` functionality that will detect the
+  ;; case where we're trying to borrow from an old pool, so that we can retry
+  ;; with the new pool.
+  [pool])
+
+(def JRubyPuppetConfig
+  "Schema defining the config map for the JRubyPuppet pooling functions.
+
+  The keys should have the following values:
+
+  * :ruby-load-path - a vector of file paths, containing the locations of puppet source code.
+
+  * :gem-home - The location that JRuby gems are stored
+
+  * :master-conf-dir - file path to puppetmaster's conf dir;
+    if not specified, will use the puppet default.
+
+  * :master-code-dir - file path to puppetmaster's code dir;
+    if not specified, will use the puppet default.
+
+  * :master-var-dir - path to the puppetmaster's var dir;
+    if not specified, will use the puppet default.
+
+  * :master-run-dir - path to the puppetmaster's run dir;
+    if not specified, will use the puppet default.
+
+  * :master-log-dir - path to the puppetmaster's log dir;
+    if not specified, will use the puppet default.
+
+  * :max-active-instances - The maximum number of JRubyPuppet instances that
+      will be pooled. If not specified, the system's
+      number of CPUs+2 will be used.
```

Comment:
This should be updated to the new algorithm, right?  (Probably need to make the same change to the "stable" branch, too, since I think this just came from there in the merge-up).